modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj

```diff
@@ -146,6 +154,90 @@
                table-schema->metabase-field-info
                set)})
 
+(def ^:private ^DateTimeFormatter bq-ts-str-formatter
+  "For some crazy reason, the TIMESTAMP String values returned from [[FieldValue]] look like `1577836800.0`, so create
+  a parser for that.  This is basically '<epoch seconds>.<millis>'."
+  (let [^DateTimeFormatterBuilder fb (doto (DateTimeFormatterBuilder.)
+                                       (.appendValue ChronoField/INSTANT_SECONDS 1 19 SignStyle/NEVER)
+                                       (.appendLiteral \.)
+                                       (.appendValue ChronoField/MILLI_OF_SECOND))]
+    (.toFormatter fb)))
+
+(defn- bq-ts-str->instant
+  "Parses the given `ts-str` (assumed to be a String representing a BigQuery TIMESTAMP value) to an [[Instant]].  See
+  [[bq-ts-str-formatter]] for more details."
+  [ts-str]
+  (let [parsed (.parse bq-ts-str-formatter ts-str)]
+    (Instant/from parsed)))
+
+;; https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#integer_type
+(def ^:private int-fingerprint-types
+  "Integral types for fingerprinting purposes (to be parsed via BigInteger)"
+  #{"INT64" "INT" "SMALLINT" "INTEGER" "BIGINT" "TINYINT" "BYTEINT"})
+
+;; https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#decimal_types
+;; https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#floating_point_types
+(def ^:private dec-fingerprint-types
+  "Decimal and floating point types for fingerprinting purposes (to be parsed via BigDecimal)"
+  #{"NUMERIC" "DECIMAL" "BIGNUMERIC" "BIGDEIMAL" "FLOAT" "FLOAT64"})
+
+(defn- fields->indexed-parser-fns
+  "For the given `fields` (instances of the [[models/Field]]), return an ordered map (by key) of parser functions that
+  will parse a String value from the [[FieldValue]] BigQuery API.  The keys of the map are the database position of the
+  respective fields (and thus, the respective position of the values within a [[FieldValueList]] as returned by the
+  BigQuery SDK list operation)."
+  [fields]
+  (reduce (fn [acc {database-type :database_type, idx :database_position}]
```

Comment:
I'm a bit hesitant about using the `database_position` that we get from syncing tables against the FieldValueList. Are we sure these will always correspond? My thinking was we could get the schema from the results and combine that with the fields we care about to figure out in the indices to select and the conversion necessary.

Reply:
We can't get the schema, unfortunately, since the SDK throws an exception if you try to (from a `TableResult` coming from the `list`)